Storage/src/StorageObject.php

```diff
@@ -672,13 +683,61 @@ public function downloadToFile($path, array $options = [])
      */
     public function downloadAsStream(array $options = [])
     {
-        return $this->connection->downloadObject(
+        // This makes sure we honour the range headers specified by the user
+        $requestedBytes = $this->getRequestedBytes($options);
+        $resultStream = Utils::streamFor(null);
+        $transcodedObj = false;
+
+        // We try to deduce if the object is a transcoded object when we receive the headers.
+        $options['restOptions']['on_headers'] = function ($response) use (&$transcodedObj) {
+            $header = $response->getHeader(StorageObject::TRANSCODED_OBJ_HEADER_KEY);
+            if (is_array($header) && in_array(StorageObject::TRANSCODED_OBJ_HEADER_VAL, $header)) {
+                $transcodedObj = true;
+            }
+        };
+
+        $options += [
+            'restOnRetryExceptionFunction' => function (\Exception $e, $attempt, &$arguments)
```

Comment:
Should we be checking if a user has set this first?

Reply:
So, the `$options += ` is supposed to tackle the case where the user has supplied it before :)